src/LinearOperators.jl

```diff
@@ -124,23 +142,37 @@ end
 
 # Constructors.
 """
-    LinearOperator(M; symmetric=false, hermitian=false)
+    LinearOperator!(Mv, Mtu, M; symmetric=false, hermitian=false)
 
-Construct a linear operator from a dense or sparse matrix.
+Construct a linear operator from a dense or sparse matrix, using `Mv` as storage space
+for `M * v` and `Mtu` as storage space for `transpose(M) * u` and `adjoint(M) * u`.
 Use the optional keyword arguments to indicate whether the operator
 is symmetric and/or hermitian.
 """
-function LinearOperator(M :: AbstractMatrix{T}; symmetric=false, hermitian=false) where T
+function LinearOperator!(Mv :: Vector{T}, Mtu :: Vector{T}, M :: AbstractMatrix{T};
+                         symmetric=false, hermitian=false) where T
   nrow, ncol = size(M)
-    Mv = Vector{T}(undef, nrow)
   prod = @closure v -> mul!(Mv, M, v)
-  Mtu = Vector{T}(undef, ncol)
   tprod = @closure u -> mul!(Mtu, transpose(M), u)
   ctprod = @closure w -> mul!(Mtu, adjoint(M), w)
   F1 = typeof(prod)
   F2 = typeof(tprod)
   F3 = typeof(ctprod)
-  LinearOperator{T,F1,F2,F3}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
+  PreallocatedLinearOperator{T,F1,F2,F3}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
+end
+
+"""
+    LinearOperator(M; symmetric=false, hermitian=false)
+
+Construct a linear operator from a dense or sparse matrix.
+Use the optional keyword arguments to indicate whether the operator
+is symmetric and/or hermitian.
+"""
+function LinearOperator(M :: AbstractMatrix{T}; symmetric=false, hermitian=false) where T
+  nrow, ncol = size(M)
+  Mv = Vector{T}(undef, nrow)
+  Mtu = Vector{T}(undef, ncol)
```

Comment:
If M is `symmetric` or `hermitian`, we don't need to allocate a `Mtu` vector.
Maybe you can do something like this?
```julia
Mv = Vector{T}(undef, nrow)
if symmetric || hermitian
  LinearOperator!(Mv, Mv, M, symmetric=symmetric, hermitian=hermitian)
else
  Mtu = Vector{T}(undef, ncol)
  LinearOperator!(Mv, Mtu, M, symmetric=symmetric, hermitian=hermitian)
end
```

Reply:
Technically, if `nrow == ncol` we can use the same memory. But should we? On this case, if `A == A'` but `A != A.'`, if we really want to differentiate, we'd like the same memory for Av and Atv, but not for Aadjv. Should we have three vectors of memory for the general case?

Reply:
I think we should limit the number of vectors allocated whenever it's possible with a `PreallocatedLinearOperator`.
* If `T` is a real type : 
1. A symmetric -> 1 vector for Av, Atv, Aadjv
2. A not symmetric -> 1 vector for Av, 1 vector for Atv, Aadjv
* if `T` is a complex type:
1. A symmetric -> 1 vector for Av, Atv, 1 vector for Aadjv
2. A hermitian -> 1 vector for Av, Aadjv, 1 vector for Atv
3. A not symmetric and not hermitian -> 1 vector for Av, 1 vector for Atv, 1 vector for Aadjv